### PR TITLE
Use GetNodeConnection to establish a connection in transaction recovery

### LINKED ISSUE #1099: Convert transaction_recovery.c to new connection framework
Should be fairly simple.

Noticed because that'd allow us to remove ReraiseRemoteError().

This is part of #873

### LINKED ISSUE #1111: Call PQclear in ReportResultError
We currently don't call PQclear in `ReportResultError` in the way that `pgfdw_report_error` in postgres_fdw does, which may lead to memory leaks in case of failure.

### PR DESCRIPTION
**Pull request by marcocitus**
I couldn't find a good way to PQclear the result in case of error, so I added the PG_TRY block as found in `pgfdw_report_error` in postgres_fdw to `ReportResultError`.

Fixes #1099 
Fixes #1111

### CODE REVIEW

File: src/backend/distributed/connection/remote_commands.c [hunk 1]
```diff
@@ -125,46 +125,68 @@ ReportConnectionError(MultiConnection *connection, int elevel)
...
 ReportResultError(MultiConnection *connection, PGresult *result, int elevel)
 {
-	char *sqlStateString = PQresultErrorField(result, PG_DIAG_SQLSTATE);
-	char *messagePrimary = PQresultErrorField(result, PG_DIAG_MESSAGE_PRIMARY);
-	char *messageDetail = PQresultErrorField(result, PG_DIAG_MESSAGE_DETAIL);
-	char *messageHint = PQresultErrorField(result, PG_DIAG_MESSAGE_HINT);
-	char *messageContext = PQresultErrorField(result, PG_DIAG_CONTEXT);
+	/* PGresult must be released before leaving this function. */
+	PG_TRY();
+	{
```
Comment (anarazel) [inline src/backend/distributed/connection/remote_commands.c]: Hm. Not sure if it'd not be better to instead check elevel?

File: src/backend/distributed/connection/remote_commands.c [hunk 2]
```diff
@@ -125,46 +125,68 @@ ReportConnectionError(MultiConnection *connection, int elevel)
...
+						 messageContext ? errcontext("%s", messageContext) : 0,
+						 errcontext("while executing command on %s:%d",
+									nodeName, nodePort)));
+	}
+	PG_CATCH();
+	{
+		if (elevel == ERROR)
 		{
-			*lastNewlineIndex = '\0';
+			PQclear(result);
```
Comment (anarazel) [inline src/backend/distributed/connection/remote_commands.c]: Why do we need the check for ERROR here? Afaics we should always do this if we throw?

File: src/backend/distributed/connection/remote_commands.c [hunk 3]
```diff
@@ -125,46 +125,68 @@ ReportConnectionError(MultiConnection *connection, int elevel)
...
 	}
+	PG_END_TRY();
 
-	ereport(elevel, (errcode(sqlState), errmsg("%s", messagePrimary),
-					 messageDetail ? errdetail("%s", messageDetail) : 0,
-					 messageHint ? errhint("%s", messageHint) : 0,
-					 messageContext ? errcontext("%s", messageContext) : 0,
-					 errcontext("while executing command on %s:%d",
-								nodeName, nodePort)));
+	if (elevel == ERROR)
```
Comment (anarazel) [inline src/backend/distributed/connection/remote_commands.c]: Isn't this unreachable code?

File: src/backend/distributed/transaction/transaction_recovery.c [hunk 4]
```diff
@@ -163,8 +165,10 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
 	MemoryContext localContext = NULL;
 	MemoryContext oldContext = NULL;
 
-	PGconn *connection = GetOrEstablishConnection(nodeName, nodePort);
-	if (connection == NULL)
+	int connectionFlags = SESSION_LIFESPAN;
```
Comment (anarazel) [inline src/backend/distributed/transaction/transaction_recovery.c]: That's somewhat weird newlines around connectionFlags here?

File: src/backend/distributed/connection/remote_commands.c (hunk 1 above)
Comment (anarazel) [inline src/backend/distributed/connection/remote_commands.c]: (that'd require copying some of the strings, so this really is optional)

File: src/backend/distributed/connection/remote_commands.c (hunk 2 above)
Comment (marcocitus) [inline src/backend/distributed/connection/remote_commands.c]: uff, brain's not working well

File: src/backend/distributed/connection/remote_commands.c (hunk 1 above)
Comment (marcocitus) [inline src/backend/distributed/connection/remote_commands.c]: Well, this is taken directly from pgfdw_report_error (on which the rest of this function also seems to be based), so it seems reasonably battle-tested.

File: src/backend/distributed/transaction/transaction_recovery.c (hunk 4 above)
Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_recovery.c]: Removed the extra newline.

File: src/backend/distributed/transaction/transaction_recovery.c [hunk 5]
```diff
@@ -225,17 +230,26 @@ RecoverWorkerTransactions(WorkerNode *workerNode)
...
+		if (querySent == 0)
 		{
-			WarnRemoteError(connection, result);
+			ReportConnectionError(connection, WARNING);
+
+			break;
+		}
+
+		result = GetRemoteCommandResult(connection, raiseInterrupts);
+		if (!IsResponseOK(result))
```
Comment (anarazel) [inline src/backend/distributed/transaction/transaction_recovery.c]: Fine with this for now, but should we also add a ExecuteRemoteCommand()?

Comment (marcocitus) [inline src/backend/distributed/transaction/transaction_recovery.c]: Might be useful. We should probably merge the existing ExecuteRemoteCommand and ExecuteCriticalRemoteCommand into one function and get rid of the former.

**Pull request merged**